File: web/src/api/taskqueue.ts
import { getHttpClient } from '@/api/http-binding';

export type QueueTaskStatus =
  | 'pending'
  | 'claimed'
  | 'running'
  | 'completed'
  | 'failed'
  | 'cancelled'
  | 'retrying'
  | string;

export type QueueTask = {
  id: string;
  idempotencyKey: string;
  workspaceId: string;
  type: string;
  payload: string;
  priority: number;
  status: QueueTaskStatus;
  nodeId: string;
  agentId: string;
  queue: string;
  attempts: number;
  maxAttempts: number;
  lastError: string;
  result: string;
  timeoutMs: number;
  createdAt: number;
  updatedAt: number;
  claimedAt?: number;
  startedAt?: number;
  completedAt?: number;
};

export type ListQueueTasksInput = {
  workspaceId?: string;
  status?: string;
  queue?: string;
  nodeId?: string;
  type?: string;
};

export type CreateQueueTaskInput = {
  idempotencyKey?: string;
  workspaceId?: string;
  type: string;
  payload: string;
  priority?: number;
  queue?: string;
  maxAttempts?: number;
  timeoutMs?: number;
};

const asNumber = (v: unknown): number => {
  const n = Number(v);
  return Number.isFinite(n) ? n : 0;
};

const mapTask = (raw: Record<string, unknown>): QueueTask => ({
  id: String(raw.id ?? ''),
  idempotencyKey: String(raw.idempotencyKey ?? ''),
  workspaceId: String(raw.workspaceId ?? ''),
  type: String(raw.type ?? ''),
  payload: String(raw.payload ?? ''),
  priority: asNumber(raw.priority),
  status: String(raw.status ?? 'pending'),
  nodeId: String(raw.nodeId ?? ''),
  agentId: String(raw.agentId ?? ''),
  queue: String(raw.queue ?? 'default'),
  attempts: asNumber(raw.attempts),
  maxAttempts: asNumber(raw.maxAttempts),
  lastError: String(raw.lastError ?? ''),
  result: String(raw.result ?? ''),
  timeoutMs: asNumber(raw.timeoutMs),
  createdAt: asNumber(raw.createdAt),
  updatedAt: asNumber(raw.updatedAt),
  claimedAt: raw.claimedAt === undefined ? undefined : asNumber(raw.claimedAt),
  startedAt: raw.startedAt === undefined ? undefined : asNumber(raw.startedAt),
  completedAt: raw.completedAt === undefined ? undefined : asNumber(raw.completedAt),
});

const queryString = (input: ListQueueTasksInput = {}) => {
  const params = new URLSearchParams();
  if (input.workspaceId) params.set('workspaceId', input.workspaceId);
  if (input.status) params.set('status', input.status);
  if (input.queue) params.set('queue', input.queue);
  if (input.nodeId) params.set('nodeId', input.nodeId);
  if (input.type) params.set('type', input.type);
  const qs = params.toString();
  return qs ? `?${qs}` : '';
};

export const listQueueTasks = async (input: ListQueueTasksInput = {}): Promise<QueueTask[]> => {
  const http = getHttpClient();
  const body = await http.get<{ items?: Record<string, unknown>[] }>(`/queue/tasks${queryString(input)}`);
  return (body.items ?? []).map(mapTask).filter((task) => task.id);
};

export const createQueueTask = async (input: CreateQueueTaskInput): Promise<QueueTask> => {
  const http = getHttpClient();
  const body = await http.post<{ task?: Record<string, unknown> }>('/queue/tasks', {
    idempotencyKey: input.idempotencyKey ?? '',
    workspaceId: input.workspaceId ?? http.workspaceId,
    type: input.type,
    payload: input.payload,
    priority: input.priority ?? 2,
    queue: input.queue ?? 'default',
    maxAttempts: input.maxAttempts ?? 3,
    timeoutMs: input.timeoutMs ?? 0,
  });
  return mapTask(body.task ?? {});
};

export const claimQueueTask = async (queue: string, nodeId: string): Promise<QueueTask> => {
  const http = getHttpClient();
  const raw = await http.post<Record<string, unknown>>('/queue/claim', { queue, nodeId });
  return mapTask(raw);
};

export const claimQueueTaskById = async (taskId: string, nodeId: string, agentId = ''): Promise<QueueTask> => {
  const http = getHttpClient();
  const raw = await http.post<Record<string, unknown>>(`/queue/tasks/${encodeURIComponent(taskId)}/claim`, { nodeId, agentId });
  return mapTask(raw);
};

export const startQueueTask = async (taskId: string, nodeId: string): Promise<QueueTask> => {
  const http = getHttpClient();
  const raw = await http.post<Record<string, unknown>>(`/queue/tasks/${encodeURIComponent(taskId)}/start`, { nodeId });
  return mapTask(raw);
};

export const ackQueueTask = async (taskId: string, nodeId: string, result = ''): Promise<QueueTask> => {
  const http = getHttpClient();
  const raw = await http.post<Record<string, unknown>>(`/queue/tasks/${encodeURIComponent(taskId)}/ack`, {
    nodeId,
    result,
  });
  return mapTask(raw);
};

export const nackQueueTask = async (taskId: string, nodeId: string, error = 'Marked failed from task map'): Promise<QueueTask> => {
  const http = getHttpClient();
  const raw = await http.post<Record<string, unknown>>(`/queue/tasks/${encodeURIComponent(taskId)}/nack`, {
    nodeId,
    error,
  });
  return mapTask(raw);
};

export const cancelQueueTask = async (taskId: string): Promise<QueueTask> => {
  const http = getHttpClient();
  const raw = await http.request<Record<string, unknown>>(`/queue/tasks/${encodeURIComponent(taskId)}`, { method: 'DELETE' });
  return mapTask(raw);
};

File: backend/go/internal/api/http/handlers/taskqueue.go
package handlers

import (
	"errors"
	"net/http"
	"strings"
	"time"

	"open-kraken/backend/go/internal/taskqueue"
)

// TaskQueueHandler serves task queue API endpoints.
type TaskQueueHandler struct {
	svc        *taskqueue.Service
	pathPrefix string // e.g. "/api/v1/queue"
}

// NewTaskQueueHandler creates a TaskQueueHandler.
func NewTaskQueueHandler(svc *taskqueue.Service, pathPrefix string) *TaskQueueHandler {
	return &TaskQueueHandler{svc: svc, pathPrefix: strings.TrimRight(pathPrefix, "/")}
}

// Handle dispatches /queue, /queue/tasks, /queue/tasks/{id}, /queue/claim, /queue/stats.
func (h *TaskQueueHandler) Handle(w http.ResponseWriter, r *http.Request) {
	sub := strings.TrimPrefix(r.URL.Path, h.pathPrefix)
	sub = strings.TrimPrefix(sub, "/")

	switch {
	case sub == "tasks" || sub == "":
		h.handleTasks(w, r)
	case sub == "claim":
		h.handleClaim(w, r)
	case sub == "stats":
		h.handleStats(w, r)
	case strings.HasPrefix(sub, "tasks/"):
		taskID := strings.TrimPrefix(sub, "tasks/")
		parts := strings.SplitN(taskID, "/", 2)
		if len(parts) == 2 {
			h.handleTaskAction(w, r, parts[0], parts[1])
		} else {
			h.handleTaskByID(w, r, parts[0])
		}
	default:
		w.WriteHeader(http.StatusNotFound)
	}
}

func (h *TaskQueueHandler) handleTasks(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		q := taskqueue.Query{
			WorkspaceID: r.URL.Query().Get("workspaceId"),
			Status:      taskqueue.TaskStatus(r.URL.Query().Get("status")),
			QueueName:   r.URL.Query().Get("queue"),
			NodeID:      r.URL.Query().Get("nodeId"),
			Type:        r.URL.Query().Get("type"),
		}
		tasks, err := h.svc.List(r.Context(), q)
		if err != nil {
			writeError(w, http.StatusInternalServerError, err)
			return
		}
		if tasks == nil {
			tasks = []taskqueue.Task{}
		}
		writeJSON(w, http.StatusOK, map[string]any{"items": marshalTasks(tasks)})

	case http.MethodPost:
		var body struct {
			IdempotencyKey string `json:"idempotencyKey"`
			WorkspaceID    string `json:"workspaceId"`
			Type           string `json:"type"`
			Payload        string `json:"payload"`
			Priority       int    `json:"priority"`
			QueueName      string `json:"queue"`
			MaxAttempts    int    `json:"maxAttempts"`
			TimeoutMs      int64  `json:"timeoutMs"`
		}
		if !decodeJSON(r, &body, w) {
			return
		}
		t := taskqueue.Task{
			IdempotencyKey: body.IdempotencyKey,
			WorkspaceID:    body.WorkspaceID,
			Type:           body.Type,
			Payload:        body.Payload,
			Priority:       taskqueue.Priority(body.Priority),
			QueueName:      body.QueueName,
			MaxAttempts:    body.MaxAttempts,
			Timeout:        time.Duration(body.TimeoutMs) * time.Millisecond,
		}
		created, err := h.svc.Enqueue(r.Context(), t)
		if err != nil {
			writeError(w, http.StatusBadRequest, err)
			return
		}
		writeJSON(w, http.StatusCreated, map[string]any{"task": marshalTask(created)})

	default:
		w.WriteHeader(http.StatusMethodNotAllowed)
	}
}

func (h *TaskQueueHandler) handleTaskByID(w http.ResponseWriter, r *http.Request, id string) {
	switch r.Method {
	case http.MethodGet:
		t, err := h.svc.Get(r.Context(), id)
		if err != nil {
			writeTaskError(w, err)
			return
		}
		writeJSON(w, http.StatusOK, marshalTask(t))

	case http.MethodDelete:
		t, err := h.svc.Cancel(r.Context(), id)
		if err != nil {
			writeTaskError(w, err)
			return
		}
		writeJSON(w, http.StatusOK, marshalTask(t))

	default:
		w.WriteHeader(http.StatusMethodNotAllowed)
	}
}

// handleTaskAction handles /queue/tasks/{id}/{action} (claim, start, ack, nack).
func (h *TaskQueueHandler) handleTaskAction(w http.ResponseWriter, r *http.Request, id, action string) {
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	var body struct {
		NodeID  string `json:"nodeId"`
		AgentID string `json:"agentId"`
		Result  string `json:"result"`
		Error   string `json:"error"`
	}
	if !decodeJSON(r, &body, w) {
		return
	}

	ctx := r.Context()
	var t taskqueue.Task
	var err error

	switch action {
	case "claim":
		t, err = h.svc.ClaimByID(ctx, id, body.NodeID, body.AgentID)
	case "start":
		t, err = h.svc.Start(ctx, id, body.NodeID)
	case "ack":
		t, err = h.svc.Ack(ctx, id, body.NodeID, body.Result)
	case "nack":
		t, err = h.svc.Nack(ctx, id, body.NodeID, body.Error)
	default:
		w.WriteHeader(http.StatusNotFound)
		return
	}

	if err != nil {
		writeTaskError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, marshalTask(t))
}

func (h *TaskQueueHandler) handleClaim(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	var body struct {
		QueueName string `json:"queue"`
		NodeID    string `json:"nodeId"`
	}
	if !decodeJSON(r, &body, w) {
		return
	}
	t, err := h.svc.Claim(r.Context(), body.QueueName, body.NodeID)
	if err != nil {
		writeTaskError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, marshalTask(t))
}

func (h *TaskQueueHandler) handleStats(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	stats, err := h.svc.Stats(r.Context(), r.URL.Query().Get("workspaceId"))
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}
	writeJSON(w, http.StatusOK, stats)
}

func marshalTask(t taskqueue.Task) map[string]any {
	m := map[string]any{
		"id":             t.ID,
		"idempotencyKey": t.IdempotencyKey,
		"workspaceId":    t.WorkspaceID,
		"type":           t.Type,
		"payload":        t.Payload,
		"priority":       int(t.Priority),
		"status":         string(t.Status),
		"nodeId":         t.NodeID,
		"agentId":        t.AgentID,
		"queue":          t.QueueName,
		"attempts":       t.Attempts,
		"maxAttempts":    t.MaxAttempts,
		"lastError":      t.LastError,
		"result":         t.Result,
		"timeoutMs":      t.Timeout.Milliseconds(),
		"createdAt":      t.CreatedAt.UnixMilli(),
		"updatedAt":      t.UpdatedAt.UnixMilli(),
	}
	if !t.ClaimedAt.IsZero() {
		m["claimedAt"] = t.ClaimedAt.UnixMilli()
	}
	if !t.StartedAt.IsZero() {
		m["startedAt"] = t.StartedAt.UnixMilli()
	}
	if !t.CompletedAt.IsZero() {
		m["completedAt"] = t.CompletedAt.UnixMilli()
	}
	return m
}

func marshalTasks(tasks []taskqueue.Task) []map[string]any {
	out := make([]map[string]any, 0, len(tasks))
	for _, t := range tasks {
		out = append(out, marshalTask(t))
	}
	return out
}

func writeTaskError(w http.ResponseWriter, err error) {
	switch err {
	case nil:
		return
	}
	switch {
	case errors.Is(err, taskqueue.ErrNotFound):
		writeError(w, http.StatusNotFound, err)
	case errors.Is(err, taskqueue.ErrInvalidTransition), errors.Is(err, taskqueue.ErrAlreadyClaimed), errors.Is(err, taskqueue.ErrNoAvailableAgent):
		writeError(w, http.StatusConflict, err)
	case errors.Is(err, taskqueue.ErrAlreadyExists):
		writeError(w, http.StatusConflict, err)
	default:
		writeError(w, http.StatusBadRequest, err)
	}
}
